server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -209,9 +246,9 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
         final int indexingOps = bulkRequest.numberOfActions();
         final long indexingBytes = bulkRequest.ramBytesUsed();
         final boolean isOnlySystem = isOnlySystem(bulkRequest, clusterService.state().metadata().getIndicesLookup(), systemIndices);
+        final String executorName = isOnlySystem ? Names.SYSTEM_WRITE : Names.WRITE;
         final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
```

Comment:
Ah wait, sorry, we need to do this check first, on the transport worker, before we start waiting for cluster states. Otherwise we might just pile up far too much work in memory waiting for the cluster recovery.

But that is a little tricky because we need to compute `isOnlySystem` to call this, and that needs the cluster to be recovered. I suggest we conservatively assume `isOnlySystem == false` if the cluster is not yet recovered.

Reply:
Sorry I don't quite follow. It seems to me that when the code reaches here, the cluster state is either recovered successfully or we had no need to wait for recover at all. So there is no more waiting after this line?

Reply:
Yes that's right, but it's the waiting _before_ this line that's a problem. Before we run this line we have the bulk request in memory but we aren't tracking it in the indexing pressure subsystem. If we receive lots of indexing requests while the cluster is recovering then we will try to hold them all in memory, rejecting none of them and will eventually just OOM.

Reply:
Makes sense. Thanks a lot for the explanation! Updated in 812154a8acf646b4c722e171d6200420838297af

Reply:
Btw, I didn't change how `isOnlySystem` is computed since the method is written with the assumption that the index may not be available yet (see [here](https://github.com/elastic/elasticsearch/pull/61655/files#r491024993)).
Also, if the metadata is not available, the ultimate default is to return `false` (when the `SystemIndices` also does not know the index).

Reply:
I had some concerns about whether `clusterState.metadata()` could be null, or `clusterState.metadata().getIndicesLookup()`, but I did some checking and I think this is ok.

Reply:
Thanks checking! I didn't think that was possible.